gstools/variogram/estimator.pyx

```diff
@@ -102,100 +152,143 @@ cdef _normalization_func choose_estimator_normalization(str estimator_type):
+    return normalization_func_vec
 
 
-def unstructured(
-    const double[:] f,
+def directional(
+    const int dim,
+    const double[:,:] f,
     const double[:] bin_edges,
-    const double[:] x,
-    const double[:] y=None,
-    const double[:] z=None,
+    const double[:,:] pos,
+    const double[:,:] direction,  # should be normed
+    const double angles_tol=M_PI/8.0,
+    const double bandwidth=-1.0,  # negative values to turn of bandwidth search
     str estimator_type='m'
 ):
-    if x.shape[0] != f.shape[0]:
-        raise ValueError('len(x) = {0} != len(f) = {1} '.
-                         format(x.shape[0], f.shape[0]))
+    if pos.shape[1] != f.shape[1]:
+        raise ValueError('len(pos) = {0} != len(f) = {1} '.
+                         format(pos.shape[1], f.shape[1]))
+
     if bin_edges.shape[0] < 2:
         raise ValueError('len(bin_edges) too small')
 
-    cdef _dist_func distance
-    # 3d
-    if z is not None:
-        if z.shape[0] != f.shape[0]:
-            raise ValueError('len(z) = {0} != len(f) = {1} '.
-                             format(z.shape[0], f.shape[0]))
-        distance = _distance_3d
-    # 2d
-    elif y is not None:
-        if y.shape[0] != f.shape[0]:
-            raise ValueError('len(y) = {0} != len(f) = {1} '.
-                             format(y.shape[0], f.shape[0]))
-        distance = _distance_2d
-    # 1d
-    else:
-        distance = _distance_1d
+    if angles_tol <= 0:
+        raise ValueError('tolerance for angle search masks must be > 0')
+
+    cdef _estimator_func estimator_func = choose_estimator_func(estimator_type)
+    cdef _normalization_func_vec normalization_func_vec = (
+        choose_estimator_normalization_vec(estimator_type)
+    )
+
+    cdef int d_max = direction.shape[0]
+    cdef int i_max = bin_edges.shape[0] - 1
+    cdef int j_max = pos.shape[1] - 1
+    cdef int k_max = pos.shape[1]
+    cdef int f_max = f.shape[0]
+
+    cdef double[:,:] variogram = np.zeros((d_max, len(bin_edges)-1))
+    cdef long[:,:] counts = np.zeros((d_max, len(bin_edges)-1), dtype=long)
+    cdef vector[double] pos1 = vector[double](dim, 0.0)
+    cdef vector[double] pos2 = vector[double](dim, 0.0)
+    cdef int i, j, k, m, d
+    cdef DTYPE_t dist
+
+    for i in prange(i_max, nogil=True):
+        for j in range(j_max):
+            for k in range(j+1, k_max):
+                dist = distance(dim, pos, j, k)
+                if dist >= bin_edges[i] and dist < bin_edges[i+1]:
+                    for d in range(d_max):
+                        if dir_test(dim, pos, dist, direction, angles_tol, bandwidth, k, j, d):
+                            for m in range(f_max):
+                                # skip no data values
+                                if not (isnan(f[m,k]) or isnan(f[m,j])):
+                                    counts[d, i] += 1
+                                    variogram[d, i] += estimator_func(f[m,k] - f[m,j])
+
```

Comment:
One could argue to incorporate a `break` here, since if a point pair matches one direction, it (maybe) shouldn't be used in another direction as well. This only happens if the angels_tol is big enough, so two directions have an overlapping search area. @LSchueler your opinion?

Reply:
Every hardcore Bayesian would probably cry "heresy!", but maybe it is exactly the intention of someone using a large tolerance to do a preliminary test without any knowledge about the main directions of the field. Then, I guess it would be helpful to use as many data points as possible, even if they overlap and are redundant.

Reply:
For speed up reasons we could check all angles between directions and if their maximum is less than  `2 * angels_tol`, we could break ;-)

Reply:
Fixed that by checking if the directions are separated. If the search bands overlap, point pairs are counted repeatedly.